sound/soc/sof/intel/hda-dsp.c

```diff
@@ -808,6 +825,32 @@ int hda_dsp_suspend(struct snd_sof_dev *sdev, u32 target_state)
 						HDA_VS_INTEL_EM2_L1SEN,
 						HDA_VS_INTEL_EM2_L1SEN);
 
+#if IS_ENABLED(CONFIG_SND_SOC_SOF_HDA)
+		/* clear rirb status */
+		snd_hdac_chip_writeb(bus, RIRBSTS, RIRB_INT_MASK);
+		/* disable CORB/RIRB */
+		snd_hdac_bus_stop_cmd_io(bus);
+#endif
+		/* clear stream status */
+		list_for_each_entry(stream, &bus->stream_list, list) {
+			sd_offset = SOF_STREAM_SD_OFFSET(stream);
+			snd_sof_dsp_write(sdev, HDA_DSP_HDA_BAR,
+					  sd_offset +
+					  SOF_HDA_ADSP_REG_CL_SD_STS,
+					  SOF_HDA_CL_DMA_SD_INT_MASK);
```

Comment:
this also not symmetrical, on resume you don't touch the stream status. Is this intentional and if yes please document it in the resume sequence.

Reply:
Yes I don't clear it on resume as there were no activity of the stream in between suspend and resume. As for documenting it - do you want me to write a comment on a resume saying that we don't clear stream status because of lack of activity during suspend?

Reply:
I was suggesting a comment such as:

"on a suspend, the stream status is cleared but the DMA is not stopped. During the resume operation, the TRIGGER_START will take care of reprogramming the stream status"

that said, I no longer understand this patch.

````c
int hda_dsp_stream_trigger(struct snd_sof_dev *sdev

	case SNDRV_PCM_TRIGGER_SUSPEND:
	case SNDRV_PCM_TRIGGER_PAUSE_PUSH:
	case SNDRV_PCM_TRIGGER_STOP:
		snd_sof_dsp_update_bits(sdev, HDA_DSP_HDA_BAR,
					sd_offset,
					SOF_HDA_SD_CTL_DMA_START |
					SOF_HDA_CL_DMA_SD_INT_MASK, 0x0);

		ret = snd_sof_dsp_read_poll_timeout(sdev, HDA_DSP_HDA_BAR,
						sd_offset, run,
						!(run &	dma_start),
						HDA_DSP_REG_POLL_INTERVAL_US,
						HDA_DSP_STREAM_RUN_TIMEOUT);

		if (ret < 0) {
			dev_err(sdev->dev,
				"error: %s: cmd %d: timeout on STREAM_SD_OFFSET read\n",
				__func__, cmd);
			return ret;
		}

		snd_sof_dsp_write(sdev, HDA_DSP_HDA_BAR, sd_offset +
				  SOF_HDA_ADSP_REG_CL_SD_STS,
				  SOF_HDA_CL_DMA_SD_INT_MASK);
````

The last part is the same as what you are doing in hda-dsp.c, so why do we need to do this in the suspend routine?

Reply:
What you miss is that we have done steps you quoted above only for streams that are **not** d0i3 capable. For those d0i3 capable, like WoV pipelines we don't call hda_dsp_stream_trigger() at all. Check this out

```
if (sdev->system_suspend_target == SOF_SUSPEND_S0IX &&
    spcm->stream[substream->stream].d0i3_compatible) {
			/*
			 * trap the event, not sending trigger stop to
			 * prevent the FW pipelines from being stopped,
			 * and mark the flag to ignore the upcoming DAPM
			 * PM events.
			 */
			spcm->stream[substream->stream].suspend_ignored = true;
			return 0;
}
```

Reply:
Well then the TRIGGER_START does not reprogram the stream either, so how can this possibly work?

Put differently, with your sequence, where is the stream status restored?

Reply:
> Put differently, with your sequence, where is the stream status restored?

Nowhere, there is no need to restore stream status.

Reply:
care to explain why and when it is required to program the stream status? 

You may be perfectly right but you have to explain your points so that the findings are shared. I shouldn't have to ask what the assumptions are.

Reply:
You clean stream status when you terminate its operation, here it is performed for s0ix entry because no stream operation can be running at that time.

Reply:
sorry, that does not explain why the stream state is not restored on S0ix exit. please re-read the question, I was asking about symmetries between entry and exit. If you clear any value on entry, and do not restore it on exit, then why did we set the value in the first place, or is the value restored somewhere else.

Reviewers multi-task and don't have the time to deep-dive into your code, so will apply simple procedural checks:
if you allocate memory, I will check when it is freed.
If you do something on suspend, I will check if the opposite action is done on resume.

 And at this point it's been 2 days without clarifications.

Reply:
Ok, I will try to explain it a little bit more. Let me use your memory allocation example - imagine we have allocated some memory for the stream and after some period of time a request to suspend this stream has arrived. Now, we don't free this memory as this stream has only been suspended and HW design doesn't require this memory region to be freed for low power state at level X. However the hardware requires that all operations of data copy shall be terminated during this particular low power state and checks status bit for any running/pending/non-completed/stalled/error transations. So what should we do to satisfy hardware constrains? Certainly STOP all transactions but also (since HW checks that) we clear any history of transactions which is stored in status registers. At this point we successful enter s0ix state...time passes and we resume from it now what we need to do? Certainly START previously stopped transactions, but hey, during suspend we also cleared status registers maybe we should also clear them here to keep "the symetry"? Well, we can but there is no point in doing that as during suspend there were no transaction (everything has been stopped) so there is nothing to clear. 
Same story in our stream status - we can clear them again on resume, sure there is nothing which stops us from doing that but IMHO it is redundant due to reasoning described above.

@plbossart, does it answer your concern now?

Reply:
@mrajwa this was a convoluted answer that does not address my specific question on stream state registers. You may clear whatever you feel necessary on suspend, but please explain why, and explain when these bits are restored. and if they are not restored please explain as well.

Reply:
@mrajwa  I also dont understand why you'd clear the stream status without stopping the DMA.

Reply:
Ohh common

> @mrajwa I also dont understand why you'd clear the stream status without stopping the DMA.

Without stopping DMA? What do you think `snd_hdac_bus_stop_cmd_io()` call does?

@plbossart I have no idea how can I explain it any better. I already said why do we have to clear it and also why we doing it again at resume is redundant. Please reread recent comments on this matter. If you feel I am wrong, fine just say why do you think so and what would you like to do differently.

Reply:
````c
/**
 * snd_hdac_bus_stop_cmd_io - clean up CORB/RIRB buffers
 * @bus: HD-audio core bus
 */
void snd_hdac_bus_stop_cmd_io(struct hdac_bus *bus)
{
	spin_lock_irq(&bus->reg_lock);
	/* disable ringbuffer DMAs */
	snd_hdac_chip_writeb(bus, RIRBCTL, 0);
	snd_hdac_chip_writeb(bus, CORBCTL, 0);
	spin_unlock_irq(&bus->reg_lock);

	hdac_wait_for_cmd_dmas(bus);

	spin_lock_irq(&bus->reg_lock);
	/* disable unsolicited responses */
	snd_hdac_chip_updatel(bus, GCTL, AZX_GCTL_UNSOL, 0);
	spin_unlock_irq(&bus->reg_lock);
}
````

what does this have to do with stream status?

Please take feedback seriously, you are gradually losing credibility and this PR will go nowhere.

Reply:
> what does this have to do with stream status?

I mentioned `snd_hdac_bus_stop_cmd_io()` to address @ranj063  suggestion that I clear stream status without stopping the DMA. As you can see DMA is stopped.

Reply:
@mrajwa I'm sorry I dont understand the connection here. Your stopping the CORB/RIRB DMA engines but dont you have to stop the DMA engine associated with the stream by wrting to SDCTL before you clear the stream status.

Reply:
I have a gut feeling this step for clearing the stream status probably has no effect on the S0ix entry at all. Have you tried with this?

Reply:
@mrajwa Hmm, I think @ranj063 raises a valid point. stop_cmd_io() only stops the CORB/RIRB transfers. Cleaning up stream status SDxSTS without stopping the stream DMA (SDxCTL) does not look correct. Either we need to stop the stream engines like in hda_dsp_ctrl_stop_chip(), or it's not needed and we don't even clear SDxSTS.

Reply:
Disabling DMA engines by writing to SDxCTL was part of version one of this patch. However during tests on versions platforms I noticed that it is enough to disable CORB/RIRB trafic in order to get clocks deasserted. I think its actually not a bad idea to add this disabling back...will update this is a moment.
However on this thread the main issue @plbossart was concerned about is the fact that during suspend I clear stream status registers but I don't do this on a resume. Is that still not understood?

Reply:
> @plbossart was concerned about is the fact that during suspend I clear stream status registers but I don't do this on a resume. Is that still not understood?

I was asking why we don't set the bits on resume, and I didn't see an answer I (or anyone) could understand.

Reply:
@plbossart, around 10 comments above I wrote:

> during suspend we also cleared status registers maybe we should also clear them here to keep "the symetry"? Well, we can but there is no point in doing that as during suspend there were no transaction (everything has been stopped) so there is nothing to clear.

Doesn't it answer the question why we don't write to stream status again?
Maybe you are confused because you see such 'symmetry' during regular suspend/resume but here commes the difference - there we reset HDA controller therefore we need to reprogram almost everything as resetting HDA controllers bring all registers to its initial values.

Reply:
So you clear the stream status on S0ix entry, and never restore it? Why is this the case?

Reply:
Yes, as there is nothing to restore - it's a 'status' register right? What would you like to restore?

Reply:
I give up.

Reply:
@plbossart I think the missing piece here might be the missing stream DMA stop and restart. The sequence should have a stop DMA followed by clearing the stream status during suspend and then restarting the DMA during resume.

Reply:
@ranj063 this really begs the question if we should not deal with all suspend/resume in platform-specific code, and not have any references to S0ix in the core. That way we could really refactor what is common and what's not, without re-inventing an S0ix code that looks the same except where different.

Reply:
@plbossart but it would be hard to not have any reference to S0ix in the core. The PCM trigger for example needs to know whether the suspend trigger is for S3 or S0ix to do the right thing. I just talked with marcin about changing the sequence here a bit to call the platform-specific trigger for S0ix to stop/start the DMA. We'll get back with the change in a bit.